pkg/controller/podautoscaler/horizontal.go

```diff
@@ -992,9 +992,9 @@ func getLongestPolicyPeriod(scalingRules *autoscalingv2.HPAScalingRules) int32 {
 
 // calculateScaleUpLimitWithScalingRules returns the maximum number of pods that could be added for the given HPAScalingRules
 func calculateScaleUpLimitWithScalingRules(currentReplicas int32, scaleEvents []timestampedScaleEvent, scalingRules *autoscalingv2.HPAScalingRules) int32 {
-	var result int32
```

Comment:
Hi, andrew. I think initialized the result as currentReplicas is just enough.

Reply:
That is not feasible
take ScaleUp for example:  currentReplicas=1
```
    scaleUp:
      policies:
      - periodSeconds: 60
        type: Percent
        value: 50
      - periodSeconds: 60
        type: Pods
        value: 2
      selectPolicy: Min
      stabilizationWindowSeconds: 300
```
 Min{1,2,3},  result=1  This is not what we expected

Reply:
Yeah, I agree with you. But there should have a default value in case of len(proposals) == 0.

Reply:
Is it like this:  var proposals []int32 = make([]int32,len(scalingRules.Policies))

Reply:
My concern is the edge cases in line 1181 when len(proposals) == 0

Reply:
ScalingRules have default values
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/v2beta2/defaults.go#L33

Reply:
That's good. Thank you for telling me that!

Reply:
@zuquan-song raises a good point. @JoshuaAndrew The issue with the code right now is that the initial `result` is not initialized correctly for both cases: `MinSelectPolicy` and `MaxSelectPolicy`. So initializing both correctly should be enough, like with this diff:
```diff
diff --git a/pkg/controller/podautoscaler/horizontal.go b/pkg/controller/podautoscaler/horizontal.go
index 950f04787f3..41231f7b727 100644
--- a/pkg/controller/podautoscaler/horizontal.go
+++ b/pkg/controller/podautoscaler/horizontal.go
@@ -998,8 +998,10 @@ func calculateScaleUpLimitWithScalingRules(currentReplicas int32, scaleEvents []
 	if *scalingRules.SelectPolicy == autoscalingv2.DisabledPolicySelect {
 		return currentReplicas // Scaling is disabled
 	} else if *scalingRules.SelectPolicy == autoscalingv2.MinPolicySelect {
+		result = math.MaxInt32
 		selectPolicyFn = min // For scaling up, the lowest change ('min' policy) produces a minimum value
 	} else {
+		result = math.MinInt32
 		selectPolicyFn = max // Use the default policy otherwise to produce a highest possible change
 	}
 	for _, policy := range scalingRules.Policies {
@@ -1018,15 +1020,17 @@ func calculateScaleUpLimitWithScalingRules(currentReplicas int32, scaleEvents []
 
 // calculateScaleDownLimitWithBehavior returns the maximum number of pods that could be deleted for the given HPAScalingRules
 func calculateScaleDownLimitWithBehaviors(currentReplicas int32, scaleEvents []timestampedScaleEvent, scalingRules *autoscalingv2.HPAScalingRules) int32 {
-	var result int32 = math.MaxInt32
+	var result int32
 	var proposed int32
 	var selectPolicyFn func(int32, int32) int32
 	if *scalingRules.SelectPolicy == autoscalingv2.DisabledPolicySelect {
 		return currentReplicas // Scaling is disabled
 	} else if *scalingRules.SelectPolicy == autoscalingv2.MinPolicySelect {
 		selectPolicyFn = max // For scaling down, the lowest change ('min' policy) produces a maximum value
+		result = math.MinInt32
 	} else {
 		selectPolicyFn = min // Use the default policy otherwise to produce a highest possible change
+		result = math.MaxInt32
 	}
 	for _, policy := range scalingRules.Policies {
 		replicasDeletedInCurrentPeriod := getReplicasChangePerPeriod(policy.PeriodSeconds, scaleEvents)
```

Reply:
@arjunrn  That's a good suggestion. I plan to adopt it